src/Core/MainLoop.js

```diff
@@ -65,10 +65,37 @@ function updateElements(context, geometryLayer, elements) {
         // and then update Debug.js:addGeometryLayerDebugFeatures
         const newElementsToUpdate = geometryLayer.update(context, geometryLayer, element);
 
-        // update attached layers
-        for (const attachedLayer of geometryLayer._attachedLayers) {
-            if (attachedLayer.ready) {
-                attachedLayer.update(context, attachedLayer, element);
+        const sub = geometryLayer.getObjectToUpdateForAttachedLayers(element);
+
+        if (sub) {
+            if (sub.element) {
+                if (__DEBUG__) {
+                    if (!(sub.element.isObject3D)) {
+                        throw new Error(`
+                            Invalid object for attached layer to update.
+                            Must be a THREE.Object and have a THREE.Material`);
+                    }
+                }
+                // update attached layers
+                for (const attachedLayer of geometryLayer._attachedLayers) {
+                    if (attachedLayer.ready) {
+                        attachedLayer.update(context, attachedLayer, sub.element, sub.parent);
+                    }
+                }
+            } else if (sub.elements) {
+                for (let i = 0; i < sub.elements.length; i++) {
```

Comment:
Use `for (let element of sub.elements)` instead.

Reply:
I dislike more and more the `for of ` syntax because it's harder to debug (lots of indirection). I'm also wondering on the perf cost of this (for the same reason: lots of indirection)...

Anyway I'll change it because for now we settled on using it everywhere.

Reply:
Yeah it seems that performance wise it is the slowest (as tested quickly here http://jsben.ch/iEYTg)

Reply:
And this test misses one of the cost: in itowns we never run native `for of` loops, but only transpiled `for of`.
For instance a simple:
```js
for (const word of words) {
  console.log(word);
}
```
becomes:
```js
"use strict";

var _iteratorNormalCompletion = true;
var _didIteratorError = false;
var _iteratorError = undefined;

try {
  for (var _iterator = words[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
    var word = _step.value;

    console.log(word);
  }
} catch (err) {
  _didIteratorError = true;
  _iteratorError = err;
} finally {
  try {
    if (!_iteratorNormalCompletion && _iterator.return) {
      _iterator.return();
    }
  } finally {
    if (_didIteratorError) {
      throw _iteratorError;
    }
  }
}
```
([see it live](https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=GYewTgBAFAxiB2BnALhA7uAJhEx1cQEoIBvAKAgjiRABsBTAOlpAHMoMxNCBuMgXyA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&sourceType=module&lineWrap=true&presets=es2015%2Creact%2Cstage-2&prettier=false&targets=&version=6.26.0&envVersion=))